File: Backend/app/pw_library/python_methods/lu_factorization_simple_elimination.py
import numpy as np


def lu_factorization(A):
    """
    Performs LU factorization on matrix A without pivoting.
    Returns matrices L and U such that A = L * U.
    """
    n = len(A)
    L = np.eye(n)  # Initialize L as an identity matrix
    U = A.copy()  # U starts as a copy of A

    for i in range(n):
        for j in range(i + 1, n):
            if U[i, i] == 0:
                raise ValueError("Zero pivot encountered; LU factorization without pivoting cannot proceed.")

            # Calculate the multiplier
            L[j, i] = U[j, i] / U[i, i]

            # Eliminate entries below the pivot
            U[j, i:] = U[j, i:] - L[j, i] * U[i, i:]

    return L, U


def forward_substitution(L, b):
    """
    Solves the equation Ly = b for y, where L is a lower triangular matrix.
    """
    n = len(b)
    y = np.zeros_like(b, dtype=float)

    for i in range(n):
        y[i] = (b[i] - np.dot(L[i, :i], y[:i])) / L[i, i]

    return y


def backward_substitution(U, y):
    """
    Solves the equation Ux = y for x, where U is an upper triangular matrix.
    """
    n = len(y)
    x = np.zeros_like(y, dtype=float)

    for i in range(n - 1, -1, -1):
        x[i] = (y[i] - np.dot(U[i, i + 1:], x[i + 1:])) / U[i, i]

    return x


def solve_lu(A, b):
    """
    Solves the linear system Ax = b using LU factorization.
    """
    # Perform LU factorization
    L, U = lu_factorization(A)

    # Solve Ly = b for y
    y = forward_substitution(L, b)

    # Solve Ux = y for x
    x = backward_substitution(U, y)

    return x, L, U

File: Backend/app/pw_library/rust_methods/lu_factorization_simple_elimination.rs
use nalgebra::{DMatrix, DVector};

// LU Factorization without pivoting
fn lu_factorization(a: &DMatrix<f64>) -> Result<(DMatrix<f64>, DMatrix<f64>), &'static str> {
    let n = a.nrows();
    let mut l = DMatrix::identity(n, n); // Initialize L as an identity matrix
    let mut u = a.clone(); // U starts as a copy of A

    for i in 0..n {
        for j in (i + 1)..n {
            if u[(i, i)] == 0.0 {
                return Err("Zero pivot encountered; LU factorization without pivoting cannot proceed.");
            }

            // Calculate the multiplier
            l[(j, i)] = u[(j, i)] / u[(i, i)];

            // Eliminate entries below the pivot
            for k in i..n {
                u[(j, k)] -= l[(j, i)] * u[(i, k)];
            }
        }
    }

    Ok((l, u))
}

// Forward substitution for Ly = b
fn forward_substitution(l: &DMatrix<f64>, b: &DVector<f64>) -> DVector<f64> {
    let n = b.len();
    let mut y = DVector::zeros(n);

    for i in 0..n {
        let sum = l.row(i).columns(0, i).dot(&y.rows(0, i));
        y[i] = (b[i] - sum) / l[(i, i)];
    }

    y
}

// Backward substitution for Ux = y
fn backward_substitution(u: &DMatrix<f64>, y: &DVector<f64>) -> DVector<f64> {
    let n = y.len();
    let mut x = DVector::zeros(n);

    for i in (0..n).rev() {
        let sum = u.row(i).columns(i + 1, n - i - 1).dot(&x.rows(i + 1, n - i - 1));
        x[i] = (y[i] - sum) / u[(i, i)];
    }

    x
}

// Solving Ax = b using LU factorization
fn solve_lu(a: &DMatrix<f64>, b: &DVector<f64>) -> Result<(DVector<f64>, DMatrix<f64>, DMatrix<f64>), &'static str> {
    // Perform LU factorization
    let (l, u) = lu_factorization(a)?;

    // Solve Ly = b for y
    let y = forward_substitution(&l, b);

    // Solve Ux = y for x
    let x = backward_substitution(&u, &y);

    Ok((x, l, u))
}
